### ISSUE TITLE
Class with only `__init_subclass__` raises `attr-defined`

### ISSUE THREAD

**Issue by miketheman** [python/mypy#15205]
Hello! I dug around the issue tracker to try and find any references to this kind of issue, but didn't find anything relevant. Apologies if this is a duplicate, but it seemed like it was unique enough hat 

**Bug Report**

When adding a type hint to a mixin class's function, an error is raised. Previously, the function was untyped so was not being evaluated (i.e. not in `--strict` mode).

**To Reproduce**
I boiled it down to a simpler example. The contents of `main.py`:

```python
class HasEvents:
    def __init_subclass__(cls, /, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.foo = []  # this type doesn't appear to matter
        return cls
    
    def record_event(self, *, some_other_arg: str):
        event = self.foo.append(some_other_arg)  # <== error here
        # do something else with `event`...
        return event
```

https://mypy-play.net/?mypy=1.2.0&python=3.11&gist=90a1e568fd9630201f21c28f94167138
https://gist.github.com/mypy-play/90a1e568fd9630201f21c28f94167138

**Actual Behavior**

```
main.py:8: error: "HasEvents" has no attribute "Event"  [attr-defined]
```

I found that changing `__init_subclass__` to `__init__` passes.

**Your Environment**

I tried a variety of Python and mypy versions from the playground to see if it was new, and went back to Python 3.7 and mypy 0.750 - so versions, flags and command line options have been isolated from the equation.

Pull request part of the greater project, where I disabled the check: [pypi/warehouse@`bc76ecf` (#13593)](https://github.com/pypi/warehouse/pull/13593/commits/bc76ecfd94b7044a60204cc9262a7a6ec149c277#diff-748c3e07c4c176cac1b73f30db918437676fdce27a7259e1c80cc35373e1453fL119-R131)

**Comment by hauntsaninja** [comment]
Hm, interesting. Feature request is to allow attribute assignments in `__init_subclass__` to define typed class attributes in a similar way to how attribute assignments in `__init__` let you defined typed instance attributes.

Might be a little tricky to add support for this, since would need to do this thing only in subclasses, because:
```
class X:
    def __init_subclass__(cls):
        cls.attr = 0
X.attr  # whoops, AttributeError
```
Looking at the original code, mypy is not going to understand dynamic creation of types, so not sure how well you could type this.

But for your current example, I'd just use a classvar annotation like so:
```
from typing import ClassVar

class HasEvents:
    foo: ClassVar[list[str]]
    
    def __init_subclass__(cls, /, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.foo = []
        return cls
    
    def record_event(self, *, some_other_arg: str):
        self.foo.append(some_other_arg)
```

**Comment by miketheman** [comment]
Thanks @hauntsaninja ! I initially reported it as a bug since that's what it felt like - glad to discover it's a yet-to-be-implemented-feature!

Adding a ClassVar annotation appears to have worked, thanks for the pointer! The original code is going a bit further and creating a dynamic type value based on the mixin class' name, so I don't know if I can easily add a specific type, but I think this is good enough for my use case right now.
